plugins/filter_aws/aws.c

```diff
@@ -370,6 +385,134 @@ static int get_vpc_metadata(struct flb_filter_aws *ctx)
     return ret;
 }
 
+/* Get EC2 instance tag keys from /latest/meta-data/tags/instance.
+ * Initializes ctx->tags_count, ctx->tag_keys and ctx->tag_keys_len.
+ */
+static int get_ec2_tag_keys(struct flb_filter_aws *ctx)
+{
+    int ret;
+    flb_sds_t tags_list = NULL;
+    size_t len = 0;
+    size_t tag_index = 0;
+    size_t tag_start = 0;
+    size_t tag_end = 0;
+    size_t characters_to_copy;
+    size_t i;
+
+    /* get a list of tag keys from the meta data server */
+    ret = get_metadata(ctx, FLB_FILTER_AWS_IMDS_INSTANCE_TAG, &tags_list, &len);
+    if (ret < 0) {
+        if (ret == -2) {
+            flb_warn("[filter_aws] tags not available in the EC2 instance metadata");
+            ctx->tags_count = 0;
+            return 0;
+        }
+        flb_sds_destroy(tags_list);
+        return -1;
+    }
+    /* count number of tag keys and allocate memory */
+    /* it assumes there always is at least one tag */
+    /* starts with 1, because '\n' is a split character */
+    ctx->tags_count = 1;
+    for (i = 0; i < len; i++) {
+        if (tags_list[i] == '\n') {
+            ctx->tags_count++;
+        }
+    }
```

Comment:
How is no tags handled?

Reply:
EC2ws (EC2 metadata server) returns 404 if there are 0 tags. I added handling 404 response status code in the `get_metadata` which in such case returns `-2`. Otherwise (`ret = 0`) there is at least 1 tag available, therefore count is initialized as 1.

I will improve my comment in the source code, because it's unclear right now.

Reply:
Here is a 'mock' for EC2ws, which might help and clarify what's the request/response format.

```python
#!/usr/bin/env python3

from http.server import BaseHTTPRequestHandler, HTTPServer


class HTTPHandler(BaseHTTPRequestHandler):
    def _send_response(self, status_code: int = 200, body: str = None):
        self.send_response(status_code)
        if body:
            self.send_header("Content-Type", "text/plain")
            self.send_header("Content-Length", len(body))
        self.end_headers()
        if body:
            self.wfile.write(body.encode("utf-8"))

    def do_GET(self):
        if self.path == "/latest/meta-data/instance-id/":
            self._send_response(body="i-0e66fc7f9809d7168")

        elif self.path == "/latest/meta-data/tags/instance":
            # If there are 0 tags, AWS returns a 404.
            # self._send_response(status_code=404)
            self._send_response(body="Name\nCUSTOMER_ID")

        elif self.path == "/latest/meta-data/tags/instance/Name":
            self._send_response(body="mwarzynski-fluentbit-dev")

        elif self.path == "/latest/meta-data/tags/instance/CUSTOMER_ID":
            self._send_response(body="70ec5c04-3a6e-11ed-a261-0242ac120002")

        else:
            self._send_response(status_code=500)


def run():
    httpd = HTTPServer(("", 8000), HTTPHandler)
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    httpd.server_close()


if __name__ == "__main__":
    run()

```

Reply:
> I will improve my comment in the source code, because it's unclear right now.

https://github.com/fluent/fluent-bit/pull/6065/commits/b185ddf3aebbdc1d7b3a9a0661d42c88bacaf389

Reply:
Conversation resolved due to lack of new activity. (Feel free to bring it back.)